lumeer-core/src/main/java/io/lumeer/core/facade/SearchFacade.java

```diff
@@ -60,17 +60,80 @@
    public List<Collection> searchCollections(Query query) {
       Set<Collection> collections = new HashSet<>();
 
-      if (!isCollectionSelectionOnly(query)) {
-         collections.addAll(getCollectionsByDocumentSearch(query));
+      if (query.getFulltext() != null && !query.getFulltext().isEmpty()) {
+         collections.addAll(getCollectionsByDocumentsSearch(query));
       }
 
-      collections.addAll(getCollectionsByCollectionSearch(query));
+      if (query.getDocumentIds() != null && !query.getDocumentIds().isEmpty()) {
+         collections.addAll(getCollectionsByDocumentsIds(query.getDocumentIds()));
+      }
+
+      if (isOnlyCollectionQuery(query) || collectionQueryIsNotEmpty(query)) {
+         collections.addAll(getCollectionsByCollectionSearch(query));
+      }
 
       return new ArrayList<>(collections);
    }
 
    public List<Document> searchDocuments(Query query) {
-      Map<String, Collection> collections = getCollections(query.getCollectionCodes());
+      Set<Document> documents = new HashSet<>();
+
+      if (query.getDocumentIds() != null && !query.getDocumentIds().isEmpty()) {
+         documents.addAll(getDocumentsByIds(query.getDocumentIds()));
+      }
+
+      if (!isOnlyDocumentsIdsQuery(query) || isEmptyQuery(query)) {
+         documents.addAll(searchDocumentsByFullText(query));
```

Comment:
Why do we want to search documents by fulltext in case of an empty query?

Reply:
Because we want to return all documents on empty query (and fulltext is not applied then).

Reply:
Oh, I see. But it is not obvious at all from the code and it looks like inefficient solution at first.